runtime/near-wallet-contract/implementation/wallet-contract/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+                    .unwrap_or_else(|_| env::panic_str("Invalid address registrar"));
+                ext_registrar::ext(account_id).with_static_gas(Gas::from_tgas(5))
+            };
+            let address = format!("0x{}", hex::encode(address));
+            address_registrar
+                .lookup(address)
+                .then(ext.address_check_callback(target, action))
+        }
+    };
+    Ok(promise)
+}
+
+fn action_to_promise(target: AccountId, action: near_action::Action) -> Result<Promise, Error> {
+    match action {
+        near_action::Action::FunctionCall(action) => Ok(Promise::new(target).function_call(
+            action.method_name,
+            action.args,
+            action.deposit,
+            action.gas,
+        )),
+        near_action::Action::Transfer(action) => Ok(Promise::new(target).transfer(action.deposit)),
+        near_action::Action::Stake(action) => {
+            Ok(Promise::new(target).stake(action.stake, action.public_key))
+        }
+        near_action::Action::AddKey(action) => match action.access_key.permission {
+            near_action::AccessKeyPermission::FullAccess => Err(Error::User(
+                UserError::UnsupportedAction(UnsupportedAction::AddFullAccessKey),
+            )),
+            near_action::AccessKeyPermission::FunctionCall(access) => Ok(Promise::new(target)
+                .add_access_key_allowance_with_nonce(
+                    action.public_key,
+                    access
+                        .allowance
+                        .and_then(Allowance::limited)
+                        .unwrap_or(Allowance::Unlimited),
+                    access.receiver_id,
+                    access.method_names.join(","),
+                    action.access_key.nonce,
+                )),
+        },
+        near_action::Action::DeleteKey(action) => {
+            Ok(Promise::new(target).delete_key(action.public_key))
+        }
+        near_action::Action::CreateAccount(_) => Err(Error::User(UserError::UnsupportedAction(
+            UnsupportedAction::CreateAccount,
+        ))),
+        near_action::Action::DeployContract(_) => Err(Error::User(UserError::UnsupportedAction(
+            UnsupportedAction::DeployContract,
+        ))),
+        near_action::Action::DeleteAccount(_) => Err(Error::User(UserError::UnsupportedAction(
+            UnsupportedAction::DeleteAccount,
+        ))),
+        near_action::Action::Delegate(_) => Err(Error::User(UserError::UnsupportedAction(
+            UnsupportedAction::Delegate,
+        ))),
+    }
+}
+
+fn create_ban_relayer_promise(current_account_id: AccountId) -> Promise {
+    let pk = env::signer_account_pk();
+    Promise::new(current_account_id)
+        .delete_key(pk)
+        .function_call_weight(
+            "ban_relayer".into(),
+            Vec::new(),
+            NearToken::from_yoctonear(0),
+            Gas::from_tgas(1),
+            GasWeight(1),
+        )
+}
+
+#[near_sdk::ext_contract(ext_registrar)]
```

Comment:
How would wallet contract know which address registrar SC it is gonna use?
Would it automatically match with the `AddressRegistrar` implemented in the `address-registrar` crate somehow (i.e. both modules deployed as a package in `Aptos` or `Sui`)?

Reply:
The account id for the registrar contract will be hardcoded into the wallet contract. This account id will be chosen in advance and be a locked account (ie without any access keys) with the registrar contract deployed on both testnet and mainnet.